Properly show detach script button when script is added via inspector

Kazuo256 opened the pull request:
Fixes #107059.

The SceneTreeDock was not tracking script changes in selected nodes in any capacity as far as I could assess. To do
that, my solution essentially connects the "script_changed" signal from selected nodes to
`SceneTreeDock::_update_script_button()` whenever the selection changes. It actually queues the update to make sure it
happens only once no matter how many nodes are selected.
    
However, only connecting that signal would leave previously selected nodes with a signal connection that should no
longer exist. To properly disconnect previously selected nodes, we have to store the list of currently selected nodes so
we can disconnect them when the selection changes.

The PR also replaces all uses of `SceneTreeDock::_update_script_button()` with the queued alternative, besides other minor minor improvements requested in the review. See commit message for more details.

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

Kazuo256 committed:
Show detach script button when added via inspector

Fixes #107059.

The SceneTreeDock was not tracking script changes in selected nodes in any capacity as far as I could assess. To do
that, my solution essentially connects the "script_changed" signal from selected nodes to
"SceneTreeDock::_update_script_button()" whenever the selection changes. It actually queues the update to make sure it
happens only once no matter how many nodes are selected.

However, only connecting that signal would leave previously selected nodes with a signal connection that should no
longer exist. To properly disconnect previously selected nodes, we have to store the list of currently selected nodes so
we can disconnect them when the selection changes.

The commit also includes some improvements to the SceneTreeDock class:

1. Remove unnecessary initialization in SceneTreeDock

This field is already initialized in the line that declares it. As such, initializing it on the constructor as well as
is redundant.

2. Queue script button updates in scene tree dock

Since we now have the option to defer the script button update and make sure it only runs once per frame, it's always
best to use the queued version of the update from a performance perspective. I'm not entirely sure if there could be any
unexpected side effects but it is a minor self-contained UI update, so it's likely a relatively safe change.

The replacement includes the bindings since it is a requirement for the other replacements in the class to work
(UndoRedo needs their method names registered in the class DB). It should be OK to remove the old non-queued bindings
too even though they are accessible in the public API because it is a "unofficial" method starting with an underscore.

KoBeWi commented on editor/scene_tree_dock.h:
```diff
@@ -135,6 +135,7 @@ class SceneTreeDock : public VBoxContainer {
 
 	EditorData *editor_data = nullptr;
 	EditorSelection *editor_selection = nullptr;
+	List<Node *> node_previous_selection;
```
```suggestion
	LocalVector<Node *> node_previous_selection;
```
No reason to use a List.
Also remember to `reserve()` before adding elements.

Kazuo256 commented:
> You can't just connect `script_changed` to `_update_script_button()`, because it will update the button for every selected node. It should happen only once. You can add something like `_queue_update_script_button()`.

Makes sense. Do you have a suggestion for where to check for queued updates? My first guess would under the `NOTIFICATION_PROCESS` case in the `_notification()` method.

Kazuo256 commented on editor/scene_tree_dock.cpp:
```diff
@@ -2911,6 +2911,18 @@ void SceneTreeDock::_selection_changed() {
 
+	// Track script changes in selected nodes
+	for (int i = 0; i < node_previous_selection.size(); i++) {
+		Node *node = node_previous_selection.get(i);
```
Done.

Kazuo256 commented:
I've implemented the requested changes. One thing I realized is that clicking the script button in the scene tree dock to *remove* the script will call `_update_script_button` twice (regardless of how many nodes are selected). One because of the callback I added to the `script_changed` signal, and another as part of the usual logic of the remove script button. Is that OK or should I also adjust the script button to queue the update?

KoBeWi commented on editor/docks/scene_tree_dock.cpp:
```diff
@@ -2898,6 +2898,18 @@ void SceneTreeDock::_update_script_button() {
+void SceneTreeDock::_queue_update_script_button() {
+	if (update_script_button_queued) {
+		return;
+	}
+
+	update_script_button_queued = true;
+
+	callable_mp(this, &SceneTreeDock::_update_script_button).call_deferred();
```
```suggestion
	}
	update_script_button_queued = true;
	callable_mp(this, &SceneTreeDock::_update_script_button).call_deferred();
```
Unnecessary empty lines.

Kazuo256 replied:
Done.

KoBeWi commented on editor/docks/scene_tree_dock.cpp:
```diff
@@ -2911,6 +2923,18 @@ void SceneTreeDock::_selection_changed() {
 		_push_item(nullptr);
 	}
 
+	// Track script changes in selected nodes
```
```suggestion
	// Track script changes in selected nodes.
```
Also this comment should be when connecting the signal I think.

Kazuo256 replied:
Done.

KoBeWi commented on editor/docks/scene_tree_dock.h:
```diff
@@ -135,6 +135,8 @@ class SceneTreeDock : public VBoxContainer {
 	EditorData *editor_data = nullptr;
 	EditorSelection *editor_selection = nullptr;
+	LocalVector<Node *> node_previous_selection;
+	bool update_script_button_queued;
```
```suggestion
	bool update_script_button_queued = false;
```

Kazuo256 commented on editor/docks/scene_tree_dock.cpp:
```diff
@@ -4653,6 +4677,7 @@ SceneTreeDock::SceneTreeDock(Node *p_scene_root, EditorSelection *p_editor_selec
 	edited_scene = nullptr;
 	editor_data = &p_editor_data;
 	editor_selection = p_editor_selection;
+	update_script_button_queued = false;
```
Done too.

Kazuo256 commented:
Replaced all `SceneTreeDock::_update_script_button()` calls with the new queued version.

I chose to divide the PR's work into three commit since they were separate atomic changes that someone might want to cherry-pick/revert/etc independently, but I can squash them together if that is preferrable.

KoBeWi commented on editor/docks/scene_tree_dock.cpp:
```diff
@@ -4632,6 +4655,7 @@ void SceneTreeDock::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("_reparent_nodes_to_paths_with_transform_and_name"), &SceneTreeDock::_reparent_nodes_to_paths_with_transform_and_name);
 
 	ClassDB::bind_method(D_METHOD("_update_script_button"), &SceneTreeDock::_update_script_button);
```
This can be removed now.

Kazuo256 replied:
Isn't there a chance existing code in games and plugins use this method since it seems to be exposed to the public API (e.g. via GDScript)?

KoBeWi replied:
Methods starting with `_` are not part of public API.

Kazuo256 commented:
> Could you squash the commits? See our PR workflow documentation for details: https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase

No problem, I'm aware of the workflow. Just to be sure, I'll reiterate my last comment, but will squash as asked:

> I chose to divide the PR's work into three commit since they were separate atomic changes that someone might want to cherry-pick/revert/etc independently, but I can squash them together if that is preferable.

Repiteo commented:
Thanks!

xuwaters commented:
When closing the scene tab, the pointers in `LocalVector<Node *> node_previous_selection` becomes dangling pointers, which causes random crash.

xuwaters commented:
Thank you, opened an issue: #108783

Kazuo256 commented:
PR to fix the crash: #108785
